### PR TITLE
in_unix: Use v1 plugin API

### PR DESCRIPTION
**Pull request by repeatedly**
Signed-off-by: Masahiro Nakagawa <repeatedly@gmail.com>

**Which issue(s) this PR fixes**: 
No issue

**What this PR does / why we need it**: 
Use v1 plugin API in `in_unix` plugin.
In addition, add `tag` parameter to follow `in_forward`.

**Docs Changes**:
Add tag parameter

**Release Note**: 
Same as title.

### CODE REVIEW

**Comment by ganmacs** [comment]
> Use v1 plugin API in in_unix plugin.
> In addition, add tag parameter to follow in_forward.

These two things might be in different PR. it's hard to review... (at least, please split them into different commits)

File: lib/fluent/plugin/in_unix.rb [hunk 1]
```diff
@@ -14,50 +14,68 @@
...
+    desc 'The path to your Unix Domain Socket.'
+    config_param :path, :string, default: Fluent::DEFAULT_SOCKET_PATH
+    desc 'The backlog of Unix Domain Socket.'
+    config_param :backlog, :integer, default: nil
+    desc "New tag instead of incoming tag"
+    config_param :tag, :string, default: nil
+
+    def configure(conf)
+      super
+      #log.warn "'unix' input is obsoleted and will be removed. Use 'forward' instead."
```
Comment (ganmacs) [inline lib/fluent/plugin/in_unix.rb:48]: is it necessary?

File: lib/fluent/plugin/in_unix.rb [hunk 2]
```diff
@@ -14,50 +14,68 @@
...
+      end
 
       super
     end
 
-    #def listen
-    #end
+    def listen
+      if File.exist?(@path)
+        File.unlink(@path)
```
Comment (ganmacs) [inline lib/fluent/plugin/in_unix.rb:69]: It might be better to print a warning log.

File: lib/fluent/plugin/in_unix.rb [hunk 3]
```diff
@@ -80,22 +98,22 @@ def run
     # }
     def on_message(msg)
       # TODO format error
-      tag = msg[0].to_s
+      tag = @tag || (msg[0].to_s)
```
Comment (ganmacs) [inline lib/fluent/plugin/in_unix.rb:101]: should it check if tag is `nil`?

File: lib/fluent/plugin/in_unix.rb [hunk 4]
```diff
@@ -80,22 +98,22 @@ def run
...
+      when Array
         # Forward
-        es = MultiEventStream.new
+        es = Fluent::MultiEventStream.new
         entries.each {|e|
           record = e[1]
           next if record.nil?
-          time = e[0]
-          time = (now ||= EventTime.now) if time.to_i == 0
+          time = convert_time(e[0])
```
Comment (ganmacs) [inline lib/fluent/plugin/in_unix.rb:116]: since in #convert_time check `time == 0`,  it's `e[0].to_i` ?

File: lib/fluent/plugin/in_unix.rb [hunk 5]
```diff
@@ -105,39 +123,42 @@ def on_message(msg)
         record = msg[2]
         return if record.nil?
 
-        time = msg[1]
-        time = EventTime.now if time.to_i == 0
+        time = convert_time(msg[1])
```
Comment (ganmacs) [inline lib/fluent/plugin/in_unix.rb:126]: `msg[1].to_i` ?

**Comment by ganmacs** [comment]
Writing tests about tag parameter would be great.

File: lib/fluent/plugin/in_unix.rb [hunk 6]
```diff
@@ -14,50 +14,68 @@
...
+    end
 
     def start
       super
 
-      @loop = Coolio::Loop.new
       @lsock = listen
-      @loop.attach(@lsock)
-      @thread = Thread.new(&method(:run))
+      event_loop_attach(@lsock)
```
Comment (cosmo0920) [inline lib/fluent/plugin/in_unix.rb:55]: :+1:

File: lib/fluent/plugin/in_unix.rb (hunk 3 above)
Comment (repeatedly) [reply lib/fluent/plugin/in_unix.rb:101]: Maybe, msg type check is needed. Will add it.

File: lib/fluent/plugin/in_unix.rb (hunk 4 above)
Comment (repeatedly) [reply lib/fluent/plugin/in_unix.rb:116]: No. v1 supports nanosecond, so Float and EventTime are also supported.
`== 0` works with these types.

File: lib/fluent/plugin/in_unix.rb (hunk 5 above)
Comment (repeatedly) [reply lib/fluent/plugin/in_unix.rb:126]: ditto

File: lib/fluent/plugin/in_unix.rb (hunk 4 above)
Comment (ganmacs) [reply lib/fluent/plugin/in_unix.rb:116]: if `e[0]` is `nil`(I'm not sure this happens), should we do it?

File: lib/fluent/plugin/in_unix.rb (hunk 1 above)
Comment (repeatedly) [reply lib/fluent/plugin/in_unix.rb:48]: Can be removed. Will do

File: lib/fluent/plugin/in_unix.rb (hunk 2 above)
Comment (repeatedly) [reply lib/fluent/plugin/in_unix.rb:69]: Hmm... conflict path is rare case. Need warning?

Comment (ganmacs) [reply lib/fluent/plugin/in_unix.rb:69]: I think so. if this file exists, it means other processes may use it.

**Comment by repeatedly** [comment]
Applied reviews.

File: lib/fluent/plugin/in_unix.rb [hunk 7]
```diff
@@ -105,39 +127,42 @@ def on_message(msg)
...
 
+    def convert_time(time)
+      case
+      when time.nil? || (time == 0)
+        Fluent::EventTime.now
+      when time === Fluent::EventTime
+        time
+      else
+        Fluent::EventTime.from_time(Time.at(time))
+      end
```
Comment (ganmacs) [inline lib/fluent/plugin/in_unix.rb:143]: [nits]
```suggestion
  case time
  when nil, 0
    Fluent::EventTime.now
  when Fluent::EventTime
    time
  else
    Fluent::EventTime.from_time(Time.at(time))
  end
```

File: test/plugin/test_in_unix.rb [hunk 8]
```diff
@@ -142,4 +142,20 @@ def test_message_json
...
+    @d = create_driver
+    @d.run(shutdown: false, timeout: 5) do
+      @d.instance.__send__(:on_message, data)
+    end
+
+    assert_equal 0, @d.events.size
+
+    logs = @d.instance.log.logs
+    assert_equal 1, logs.select { |line|
+      line =~ / \[warn\]: incoming data is broken: msg=#{data.inspect}/
```
Comment (ganmacs) [inline test/plugin/test_in_unix.rb:178]: [nits] match?

**Pull request merged**